ext/ddtrace_profiling_native_extension/collectors_discrete_dynamic_sampler.c

```diff
@@ -190,16 +214,22 @@ static void maybe_readjust(discrete_dynamic_sampler *sampler, long now) {
   //   then probability will be > 1 (but we should clamp to 1 since probabilities higher than 1 don't make sense).
   // * If app is eventing a lot or our sampling overhead is big, then as time_to_sample_all_events_ns grows, sampling_probability will
   //   tend to 0.
-  long working_window_time_ns = long_max_of(0, window_time_ns - sampling_window_time_ns);
-  double max_allowed_time_for_sampling_ns = target_sampling_time_ns;
-  long time_to_sample_all_events_ns = sampler->events_per_ns * working_window_time_ns * sampler->sampling_time_ns;
-  if (max_allowed_time_for_sampling_ns == 0) {
-    // if we aren't allowed any sampling time at all, probability has to be 0
+  //
+  // In fact, we can simplify the equation further since the `window_time_ns` components cancel each other out:
+  //
+  //   sampling_probability = (target_overhead / 100) / (events_per_ns * sampling_time_ns * (1 - target_overhead / 100))
+  //                        = max_sampling_overhead / avg_sampling_overhead
+
+  double max_sampling_overhead = target_overhead / 100.;
+  double avg_sampling_overhead = sampler->events_per_ns * sampler->sampling_time_ns * (1 - max_sampling_overhead);
+
+  if (max_sampling_overhead == 0) {
+    // if we aren't allowed any sampling overhead at all, probability has to be 0
     sampler->sampling_probability = 0;
   } else {
     // otherwise apply the formula described above (protecting against div by 0)
-    sampler->sampling_probability = time_to_sample_all_events_ns == 0 ? 1. :
-      double_min_of(1., max_allowed_time_for_sampling_ns / time_to_sample_all_events_ns);
+    sampler->sampling_probability = avg_sampling_overhead == 0 ? 1. :
+      double_min_of(1., max_sampling_overhead / avg_sampling_overhead);
   }
```

Comment:
Are we missing the "protecting against div by 0" bit here? I'll admit it's a pretty situation to have `events_per_ns` or `sampling_time_ns` or `(1 - max_sampling_overhead)` be zero, but I can't convince myself exactly that they can't ever go there...?

Reply:
Ergh, on a closer look, it's there, I don't know how I missed it, please ignore temporary insanity.